dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp

```diff
@@ -101,80 +107,94 @@ PageDirectoryPtr PageDirectoryFactory::createFromEdit(String storage_name, FileP
 
 void PageDirectoryFactory::loadEdit(const PageDirectoryPtr & dir, const PageEntriesEdit & edit)
 {
-    PageDirectory::MVCCMapType & mvcc_table_directory = dir->mvcc_table_directory;
-
     for (const auto & r : edit.getRecords())
     {
-        if (auto it = max_apply_page_ids.find(r.page_id.high); it == max_apply_page_ids.end())
-        {
-            max_apply_page_ids[r.page_id.high] = r.page_id.low;
-        }
-        else
-        {
-            it->second = std::max(it->second, r.page_id.low);
-        }
-
         if (max_applied_ver < r.version)
             max_applied_ver = r.version;
         max_applied_page_id = std::max(r.page_id, max_applied_page_id);
 
-        auto [iter, created] = mvcc_table_directory.insert(std::make_pair(r.page_id, nullptr));
-        if (created)
+        // We can not avoid page id from being reused under some corner situation. Try to do gcInMemEntries
+        // and apply again to resolve the error.
+        if (bool ok = applyRecord(dir, r, /*throw_on_error*/ false); unlikely(!ok))
         {
-            iter->second = std::make_shared<VersionedPageEntries>();
+            dir->gcInMemEntries();
```

Comment:
If there is too much multiplexing, will gcInMemEntries be triggered too many times ?
Maybe we can add an interface name `gcInMemEntries(page_id)`, or add a TODO here ?

Reply:
```
PUT_EXT 100
REF 150 -> 100
DEL 100
REF 100 -> 40  <- exception with PS_DIR_APPLY_INVALID_STATUS thrown here
```

We can not only run GC on id==100 to get things right.